src/parser_urdf.cc

```diff
@@ -2661,56 +2670,135 @@ void URDF2SDF::ListSDFExtensions(const std::string &_reference)
+    errorStream.str(std::string());
+
     if (!_link->child_links.empty())
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, ["
-             << static_cast<int>(_link->child_links.size())
-             << "] children links ignored.\n";
+      errorStream << "urdf2sdf: ["
+                  << static_cast<int>(_link->child_links.size())
+                  << "] child links ignored.";
+      zeroMassErrors.emplace_back(
+          ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+      errorStream.str(std::string());
     }
 
     if (!_link->child_joints.empty())
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, ["
-             << static_cast<int>(_link->child_links.size())
-             << "] children joints ignored.\n";
+      errorStream << "urdf2sdf: ["
+                  << static_cast<int>(_link->child_joints.size())
+                  << "] child joints ignored.";
+      zeroMassErrors.emplace_back(
+          ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+      errorStream.str(std::string());
     }
 
     if (_link->parent_joint)
     {
-      sdfdbg << "urdf2sdf: link[" << _link->name
-             << "] has no inertia, "
-             << "parent joint [" << _link->parent_joint->name
-             << "] ignored.\n";
+      errorStream << "urdf2sdf: parent joint["
+                  << _link->parent_joint->name
+                  << "] ignored.";
+      zeroMassErrors.emplace_back(
+          ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+      errorStream.str(std::string());
     }
 
-    sdfdbg << "urdf2sdf: link[" << _link->name
-           << "] has no inertia, not modeled in sdf\n";
-    return;
+    errorStream << "urdf2sdf: link[" << _link->name
+                << "] is not modeled in sdf.";
+    zeroMassErrors.emplace_back(
+        ErrorCode::LINK_INERTIA_INVALID, errorStream.str());
+
+    // fail here if conversion is not allowed
+    if (!_parserConfig.URDFConvertLinkWithNoMassToFrame())
+    {
+      sdfwarn << zeroMassErrors;
+      return;
+    }
   }
 
   // create <body:...> block for non fixed joint attached bodies
   if ((_link->getParent() && _link->getParent()->name == "world") ||
       !g_reduceFixedJoints ||
-      (!_link->parent_joint ||
-       !FixedJointShouldBeReduced(_link->parent_joint)))
+      (!_link->parent_joint || !FixedJointShouldBeReduced(_link->parent_joint)))
   {
-    CreateLink(_root, _link, gz::math::Pose3d::Zero);
+    // convert links with zero mass to frames if needed
+    if (linkHasZeroMass && _parserConfig.URDFConvertLinkWithNoMassToFrame())
+    {
+      // This conversion can only happen if the parent joint is fixed, and is
+      // not to be reduced.
+      const bool canBeConvertedIntoFrame =
+          _link->parent_joint &&
+          _link->parent_joint->type == urdf::Joint::FIXED &&
+          (!FixedJointShouldBeReduced(_link->parent_joint) ||
```

Comment:
I noticed that if `disableFixedJointLumping = true` and we're creating a frame, we'll lose any other elements in `<link>` (eg. `<visual>`).

Here's an example based on case 2 of the tutorial:

<details><summary>test_robot.urdf</summary>

```xml
<robot name='test_robot'>
  <link name='link1'>
    <inertial>
      <mass value='0.1' />
      <origin rpy='1.570796326794895 0 0' xyz='0.123456789123456 0 0.0' />
      <inertia ixx='0.01' ixy='0' ixz='0' iyy='0.01' iyz='0' izz='0.01' />
    </inertial>
  </link>
  <link name='link2'> <!-- zero mass link -->
    <visual>
      <origin rpy="3 -0 0" xyz="0 0 0"/>
      <geometry>
        <sphere radius="2.0"/>
      </geometry>
    </visual>
  </link>
  <joint name='joint1_2' type='fixed'>
    <parent link='link1' />
    <child  link='link2' />
    <origin xyz='0.0 0.0 0.0' rpy='0.0 0.0 1.57'/>
  </joint>
  <link name='link3'>
    <inertial>
      <mass value='0.1' />
      <origin rpy='1.570796326794895 0 0' xyz='0.123456789123456 0 0.0' />
      <inertia ixx='0.01' ixy='0' ixz='0' iyy='0.01' iyz='0' izz='0.01' />
    </inertial>
  </link>
  <joint name='joint2_3' type='continuous'>
    <parent link='link2' />
    <child  link='link3' />
    <origin xyz='0.0 0.0 0.0' rpy='0.0 0.0 1.57'/>
  </joint>
  <gazebo reference='joint1_2'>
    <disableFixedJointLumping>true</disableFixedJointLumping>
  </gazebo>
</robot>
```
</details>

The current output is 

<details><summary>SDFormat output</summary>

```xml
<sdf version='1.9'>
  <model name='test_robot'>
    <link name='link1'>
      <inertial>
        <pose>0.123456789123456 0 0 1.570796326794895 0 0</pose>
        <mass>0.10000000000000001</mass>
        <inertia>
          <ixx>0.01</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.01</iyy>
          <iyz>0</iyz>
          <izz>0.01</izz>
        </inertia>
      </inertial>
    </link>
    <frame name='link2' attached_to='link1'>
      <pose relative_to='link1'>0 0 0 0 0 1.57</pose>
    </frame>
    <frame name='joint1_2' attached_to='link2'>
      <pose relative_to='link1'>0 0 0 0 0 1.57</pose>
    </frame>
    <joint name='joint2_3' type='revolute'>
      <pose relative_to='link2'>0 0 0 0 0 1.57</pose>
      <parent>link2</parent>
      <child>link3</child>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <lower>-10000000000000000</lower>
          <upper>10000000000000000</upper>
        </limit>
        <dynamics>
          <spring_reference>0</spring_reference>
          <spring_stiffness>0</spring_stiffness>
        </dynamics>
      </axis>
    </joint>
    <link name='link3'>
      <pose relative_to='joint2_3'>0 0 0 0 0 0</pose>
      <inertial>
        <pose>0.123456789123456 0 0 1.570796326794895 0 0</pose>
        <mass>0.10000000000000001</mass>
        <inertia>
          <ixx>0.01</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.01</iyy>
          <iyz>0</iyz>
          <izz>0.01</izz>
        </inertia>
      </inertial>
    </link>
  </model>
</sdf>
```

</details>

You'll notice there is no `<visual>` in the generated SDFormat file. I think we need to discuss this further, but the best option, as @traversaro alluded to, might be to warn the user if `disableFixedJointLumping = true` and any of the parent or child links have zero mass. Also, in the SDFormat output above, `joint1_2` is no longer a joint which contradicts the `disableFixedJointLumping` setting. I guess one possible solution is to warn the user and ignore the `disableFixedJointLumping` setting, i.e, do proper joint lumping, if either of the links have zero mass.

Reply:
that is certainly a dilemma, I'm beginning to see that the scope of any conversion to frame is getting narrower. Shall we reduce the scope of this PR to just:
* promote `sdfdbg` messages to `sdfwarn` for any zero mass link related errors or logs
* more explicit messages and possible solutions
* use `> 0` for mass requirement, instead of `gz::math::equals` and the epsilon of `1e-6`

In the case of fixed parent joint, joint lumping disabled, we can emit a warning similar to the case of fixed child joint, joint lumping disabled, where we suggest users remove the `disableFixedJointLumping` tag and allow the joint lumping mechanism to handle it, and ignore everything below the fixed joint in the meantime.

What do you think?

Reply:
Yes, I think that's our best option. I think we should also mention in the warning that if they want to disable joint lumping, the associated links must have non-zero mass and a valid `<inertial>` element.

I do think this PR fixes a bug (see https://github.com/gazebosim/sdf_tutorials/pull/87/files#r1128571697), so we should make sure to not regress when we reduce the scope of the PR. Hopefully the tests will catch it if we do 🤞🏾.